source/kraken/apply_disruption.cpp

```diff
@@ -387,34 +402,63 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+                continue;
+            }
+            const auto begin = impacted_vj.impacted_ranks.begin();
+            const auto end = impacted_vj.impacted_ranks.rbegin();
+
             nt::VehicleJourney* vj = vj_iterator->second;
             auto& new_vp = impacted_vj.new_vp;
 
             // We keep only stop time before to touch the first impacted stop time
-            bool first_sa = false;
-            for (const auto& st : vj->stop_time_list) {
-                // We need to get the associated base stop_time to compare its rank
-                const auto base_st = st.get_base_stop_time();
-                // stop is ignored if its stop_point is not in impacted_stops
-                // if we don't find an associated base we keep it
-                if (base_st && impacted_vj.impacted_ranks.count(base_st->order())) {
-                    LOG4CPLUS_TRACE(log, "Ignoring stop " << st.stop_point->uri << "on " << vj->uri);
-                    if (first_sa == true) {
-                        break;
+            if (impact->severity->effect == nt::disruption::Effect::REDUCED_SERVICE) {
+                for (const auto& st : vj->stop_time_list) {
+                    // We need to get the associated base stop_time to compare its rank
+                    const auto base_st = st.get_base_stop_time();
+                    if (base_st == nullptr) {
+                        LOG4CPLUS_TRACE(log, "Ignoring stop " << st.stop_point->uri << "on " << vj->uri);
+                        continue;
+                    }
+                    if (impacted_vj.impacted_ranks.count(base_st->order())) {
+                        if (*begin == base_st->order() && rs.is_start_stop(base_st->stop_point->stop_area->uri)
+                            && rs.is_blocked_start_point()) {
+                            continue;
+                        }
+                        const auto& last_stop_time = vj->stop_time_list.back();
+                        if (*end == base_st->order() && rs.is_end_stop(base_st->stop_point->stop_area->uri)
+                            && st.order() == last_stop_time.order() && rs.is_blocked_end_point()) {
+                            continue;
+                        }
+                        if (*begin != base_st->order() && *end != base_st->order()) {
+                            continue;
+                        }
                     }
-                    first_sa = true;
+                    nt::StopTime new_st = clone_stop_time_and_shift(st, vj);
+                    // A stop_time created by realtime should be initialized with skipped_stop = false
+                    new_st.set_skipped_stop(false);
+                    new_stop_times.push_back(std::move(new_st));
                 }
-                nt::StopTime new_st = st.clone();
-                if (first_sa == true) {
-                    new_st.set_pick_up_allowed(false);
+            } else {
+                for (const auto& st : vj->stop_time_list) {
```

Comment:
```suggestion
            } else if impact->severity->effect == nt::disruption::Effect::NO_SERVICE{
                // we remove all stop_times of vj after impacted_vj.impacted_ranks.begin()
                for (const auto& st : vj->stop_time_list) {
```

Reply:
If we get here, we only have two cases, NO_SERVICE and  REDUCED_SERVICE, [See](https://github.com/hove-io/navitia/pull/3756/files/3093b58029435ac2209f7ad94f075516ae6aa6a6#diff-5f2af7c09380355e531457e7fb7b74889626635dbfa09268a430b6808f464af1L369)